File: crates/sunday-contracts/src/service.rs
use serde::{Deserialize, Serialize};

use crate::common::default_schema_version;
use crate::song::SongRef;

/// Lifecycle of a planned service. Superset of Plan + Stage local states.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ServiceState {
    Draft,
    Published,
    InProgress,
    Played,
    Archived,
}

/// Canonical running-order item kind — a superset both Plan and Stage map onto.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ServiceItemKind {
    Song,
    Scripture,
    Sermon,
    Reading,
    Prayer,
    Offering,
    Announcement,
    Welcome,
    Response,
    Media,
    Gap,
    Custom,
}

/// A reference to a planned service. Plan is the master; Stage presents it and
/// Rec associates a recording with it. `starts_at` is ISO 8601 UTC.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ServiceRef {
    #[serde(default = "default_schema_version")]
    pub schema_version: u32,
    pub id: String,
    pub church_id: String,
    pub name: String,
    pub starts_at: String,
    pub state: ServiceState,
    pub was_streamed: bool,
    pub notes: Option<String>,
}

/// One row of a service's running order.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SetlistItem {
    pub position: u32,
    pub kind: ServiceItemKind,
    pub title: Option<String>,
    pub song_ref: Option<SongRef>,
    pub scripture_ref: Option<String>,
    pub key_override: Option<String>,
    pub duration_min: Option<i64>,
    pub notes: Option<String>,
}

/// A service plus its ordered items — the unit that flows Plan → Stage.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ServicePlan {
    #[serde(default = "default_schema_version")]
    pub schema_version: u32,
    pub service: ServiceRef,
    pub items: Vec<SetlistItem>,
}

File: packages/contracts/src/service.ts
import { z } from "zod";

import { nullableField, schemaVersionField } from "./common.js";
import { SongRef } from "./song.js";

/** Lifecycle of a planned service. Superset of Plan + Stage local states. */
export const ServiceState = z.enum([
  "draft",
  "published",
  "in_progress",
  "played",
  "archived",
]);
export type ServiceState = z.infer<typeof ServiceState>;

/**
 * Canonical running-order item kind. A *superset* of the kinds each app uses
 * locally — Plan (welcome/worship_set/scripture/sermon/response/closing/...) and
 * Stage (song/scripture/custom_deck/video/announcement/gap) both map onto this.
 * Anything unmapped becomes `custom`.
 */
export const ServiceItemKind = z.enum([
  "song",
  "scripture",
  "sermon",
  "reading",
  "prayer",
  "offering",
  "announcement",
  "welcome",
  "response",
  "media",
  "gap",
  "custom",
]);
export type ServiceItemKind = z.infer<typeof ServiceItemKind>;

/**
 * A reference to a planned service. Plan is the master; Stage consumes it for
 * presentation and Rec associates a recording with it. `starts_at` is an ISO
 * 8601 UTC timestamp.
 */
export const ServiceRef = z.object({
  schema_version: schemaVersionField,
  id: z.string().uuid(),
  church_id: z.string().uuid(),
  name: z.string().min(1).max(300),
  starts_at: z.string(),
  state: ServiceState,
  was_streamed: z.boolean(),
  notes: nullableField(z.string()),
});
export type ServiceRef = z.infer<typeof ServiceRef>;

/** One row of a service's running order. */
export const SetlistItem = z.object({
  position: z.number().int().min(0),
  kind: ServiceItemKind,
  title: nullableField(z.string()),
  song_ref: nullableField(SongRef),
  scripture_ref: nullableField(z.string()),
  key_override: nullableField(z.string().max(8)),
  duration_min: nullableField(z.number().int().min(0)),
  notes: nullableField(z.string()),
});
export type SetlistItem = z.infer<typeof SetlistItem>;

/** A service plus its ordered items — the unit that flows Plan → Stage. */
export const ServicePlan = z.object({
  schema_version: schemaVersionField,
  service: ServiceRef,
  items: z.array(SetlistItem),
});
export type ServicePlan = z.infer<typeof ServicePlan>;
